### PR TITLE
Allow Resource.Fingerprint to be the zero value.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**

### CODE REVIEW

File: resource/resource.go [hunk 1]
```diff
@@ -39,8 +39,14 @@ func (res Resource) Validate() error {
 	}
 	// TODO(ericsnow) Ensure Revision is 0 for OriginUpload?
 
-	if err := res.Fingerprint.Validate(); err != nil {
-		return errors.Annotate(err, "bad fingerprint")
+	if res.Fingerprint.IsZero() {
+		if res.Size > 0 {
+			return errors.NewNotValid(nil, "missing fingerprint; sized resources must have one")
```
Comment (natefinch) [inline resource/resource.go]: error text is kind of awkward.  0 is a size ;)  

I think it's sufficient to say "missing fingerprint".

File: resource/fingerprint.go [hunk 2]
```diff
@@ -47,11 +47,11 @@ func ParseFingerprint(raw string) (Fingerprint, error) {
 }
 
 // GenerateFingerprint returns the fingerprint for the provided data.
-func GenerateFingerprint(data io.Reader) (Fingerprint, error) {
+func GenerateFingerprint(reader io.Reader) (Fingerprint, error) {
```
Comment (natefinch) [inline resource/fingerprint.go]: traditionally r, but I'm not too picky.

**Comment by natefinch** [comment]
I would add a test for GenerateFingerprint(nil), and ensure it validates correctly.

**Comment by natefinch** [comment]
Otherwise, LGTM.

**Comment by ericsnowcurrently** [comment]
$$merge$$

**Comment by ericsnowcurrently** [comment]
$$try-again$$
